manifest/main.js

```diff
@@ -0,0 +1,127 @@
+}
+
+
+main(parseArgs());
+
+
+async function main(args) {
+
+  // load existing resources from disk, if requested (useful if we update one source at a time)
+  const resources = args.input ?
+    JSON.parse(fs.readFileSync(args.input, {encoding: 'utf8', flag: 'r'})) :
+    {};
+
+  for (const src of DATA_SOURCES) {
+    if (args.restrict_sources && !args.restrict_sources.includes(src.name)) {
+      continue
+    }
+
+    if (src.type !== 's3Inventory') {
+      logger.error(`Data source provided for a type we don't yet handle (${src.type})`)
+    }
+
+    /* Before we fetch the (updated) data and add it into the overall collection of resources,
+    delete any existing data from this data source */
+    for (const resourceType of Object.keys(resources)) {
+      for (const [id, resource] of Object.entries(resources[resourceType])) {
+        if (src.removeResource(resourceType, id, resource)) {
+          delete resources[resourceType][id];
+        }
+      }
+    }
+
+    const groupedObjects = (await parseInventory(
+      {bucket: src.bucket, prefix: src.prefix, name: src.name, local: args.local}
+    ))
+      .map(src.categorise)
+      .filter((item) => !!item)
+      // Collect together all items ("files") based on their assigned resourceType & id
+      .reduce((store, item) => {
+        const {resourceType, id} = item;
+        if (!store[resourceType]) store[resourceType] = {};
+        if (!store[resourceType][id]) store[resourceType][id] = [];
+        store[resourceType][id].push(item);
+        return store;
+      }, {})
+
+    // Turn those groups of files (each with the same ID) into a resource object, including
+    // versions etc and add the object into the main `resources` data structure.
+    for (const resourceType of Object.keys(groupedObjects)) {
+      const resourcesById = Object.fromEntries(
+        Object.entries(groupedObjects[resourceType])
+          .map(([id, items]) => [id, src.createResource(resourceType, id, items)])
+          .filter((resource) => !!resource)
+      )
+      merge(resources, resourceType, resourcesById);
+    }
+  }
+  
+  let output = JSON.stringify(resources, null, args.indent ? 2 : null);
+  if (args.gzip) {
+    output = await gzip(output)
+  }
+  fs.writeFileSync(args.output, output);
+}
+
+
+
+/**
+ * Merge the provided {resourceType, data} into the overall collection of resources
+ */
+function merge(resources, resourceType, data) {
+  if (!resources.hasOwnProperty(resourceType)) { // eslint-disable-line no-prototype-builtins
+    resources[resourceType] = {};  
+  }
+  Object.entries(data).forEach(([id, resource]) => {
+    if (resources[resourceType][id]) {
+      // The provided resource object is going to overwrite an existing one - this should be fatal
+      // as it indicates a collision of ID
+      logger.fatal(`Resource type ${resourceType} has an existing resource for ID ${id}`)
+    }
+    resources[resourceType][id] = resource;
+  });
+}
```

Comment:
Is the only reason for this merge function to check for collisions?

This question got me thinking more broadly, and ISTM that instead of keying resources by an `id` that's global within a namespace (`resourceType`), it may be a bit easier in the code if they're instead keyed by `(source, id)` where `id` is instead the within-source id/name.

Reply:
The merge function merges resources from a particular data-source into the overall `resources` structure.

The collision checking was along the lines of an `assert` expression as it represented an error in the indexer code.